### ISSUE TITLE
Implement cache busting/versioning for appropriate sections of the react app

### ISSUE THREAD

**Issue by pete-usds** [CDCgov/prime-central#78]
As a user interfacing with the Data Input App, my browser needs to automatically pick up new versions of code as they are released.

## Feature Request

- [ ] Decide on the way we want to implement versioning and cache busting.
- [ ] Implement it
- [ ] automate incrementing etc. as part of the CICD pipeline into master

**Comment by pete-usds** [comment]
May I suggest what is literally the first google result for "react app cache busting semver" as a starting point for discussion: https://dev.to/flexdinesh/cache-busting-a-react-app-22lk

**Comment by willusds** [comment]
Potential for Dave Methvin

**Comment by dmethvin-gov** [comment]
I'd like to get a better idea of the scope of functionality required here. For example, are we just checking when a user "opens the app" by navigating to it in the browser? Are we forcing it always to do a full reload of the multi-megabyte app each time they start it, bypassing the performance of the browser cache? Do we want to do some sort of active polling of the server so that if they leave it open in the browser for days we can still detect new versions and let them know?

So, create-react-app has [a support file for Service Workers](https://github.com/CDCgov/prime-data-input-client/blob/af2cb80a2e9942e952850e342ad9713388322956/frontend/src/serviceWorker.js) in order to provide offline operation and there are some docs in that file about it. There are pros and cons about it, the biggest con is that when there is a new version, the user _must_ exit all active tabs in order to get it to refresh on the _next_ load. The pro is that the user doesn't have to wait for the old version to load because it's cached. You can also build offline request queueing and such into the service worker so people can continue to work offline, but obviously also need to handle the batching of requests once a connection is reestablished.

**Comment by pete-usds** [comment]
I don't think we want to bypass caching completely - initially we were going to do that just to make things easy but then were like why not do this properly?

I'd say for prototype we probably just need something as simple as being sure that a reload in browser picks up code changes after a deploy (it doesn't currently, or didn't when we created this ticket), and we can tell people during testing/pilot to just hit reload when new changes are deployed.

Polling to pick up changes soon would be great longer term, I think, though.

**Comment by aliciabeckett-gov** [comment]
Is this user story- "As a user of the input app I want to be using the latest version of the app in my browser, so that I always have the latest and greatest functionality available to me"?

**Comment by aliciabeckett-gov** [comment]
Does this need to be in this sprint, or should it be a part of sprint 4?

**Comment by dmethvin-gov** [comment]
Okay, the user story statement makes it clearer. 👍 I've updated the title.

On the technical side we can implement an update check with service workers, or just by polling the server to see if the version has changed on some reasonable interval. 

When the version has changed, what should the UX be here? There are several options, such as a modal, a banner at the top of the page, an alert embedded somewhere on one particular screen such as the home screen, etc. Keep in mind the user may be in the middle of doing something when we detect that an update occurred, and may not want to be interrupted right then to refresh.

**Comment by willusds** [comment]
Potentially add service workers for this?

**Comment by willusds** [comment]
Send notification to the client for each version to "There's a new app, please refresh"

**Comment by dmethvin-gov** [comment]
During grooming we discussed and kicked to a later milestone TBD. 

Just as another data point, Assessment Generator releases once a week and we don't have any in-app new notification at all. If the server and network is configured correctly (i.e., not horribly broken) there is no need for cache busting, the user will get the most recent version when they open a new browser tab. Current messaging should tell them to _not_ leave the browser tab open across multiple days.

**Comment by pete-usds** [comment]
We'll probably continue to do multiple deploys a day through dev and pilot, but we can just get in the habit of telling people to force refresh all the time I guess.

**Comment by benwarfield-usds** [comment]
I'm not sure the server configuration isn't horribly broken...￼
